File: solutions/r/sieve/2/sieve.R
sieve <- function(limit) {
  if (limit < 2) return(c())
  
  is_prime <- rep(TRUE, limit)
  is_prime[1] <- FALSE
  
  for (num in 2:sqrt(limit)) {
    # Another optimalization: first multiple that needs to be checked is num^2 
    # Learned from ErikSchierboom
    # Second check gets rid of the issue with limit=2
    if (is_prime[num] && num^2 < limit) {
      is_prime[seq(num^2, limit, num)] <- FALSE
    }
  }

  which(is_prime)
}

File: solutions/typescript/sieve/2/sieve.ts
export function primes(limit: number): number[] {
  if (limit < 2) return [];

  let potentialPrimes = Array.from({ length: limit - 1 }, (_, i) => i + 2);

  for (let i = 2; i <= Math.sqrt(limit); i++) {
    potentialPrimes = potentialPrimes.filter(
      (num) => num === i || num % i !== 0
    );
  }

  return potentialPrimes;
}